protocol/x/ratelimit/keeper/keeper.go

```diff
@@ -218,6 +221,159 @@ func (k Keeper) SetDenomCapacity(
 	}
 }
 
+// UpdateCapacityEndBlocker is called during the EndBlocker to update the capacity for all limit params.
+func (k Keeper) UpdateCapacityEndBlocker(
+	ctx sdk.Context,
+) {
+	// Iterate through all the limit params in state.
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), []byte(types.LimitParamsKeyPrefix))
+	iterator := storetypes.KVStorePrefixIterator(store, []byte{})
+
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var limitParams types.LimitParams
+		k.cdc.MustUnmarshal(iterator.Value(), &limitParams)
+		k.UpdateCapacityForLimitParams(ctx, limitParams)
+	}
+}
+
+// UpdateCapacityForLimitParams calculates current baseline for a denom and recovers some amount of capacity
+// towards baseline.
+// Assumes that the `LimitParams` and corresponding `DenomCapacity` both exist in state.
+// Detailed math for calculating the updated capacity:
+//
+//	`baseline = max(baseline_minimum, baseline_tvl_ppm * tvl)`
+//	`capacity_diff = max(baseline, capacity-baseline) * (time_since_last_block / period)`
+//
+// This is basically saying that the capacity returns to the baseline over the course of the `period`.
+// Usually in a linear way, but if the `capacity` is more than twice the `baseline`, then in an exponential way.
+//
+//	`capacity =`
+//	    if `abs(capacity - baseline) < capacity_diff` then `capacity = baseline`
+//	    else if `capacity < baseline` then `capacity += capacity_diff`
+//	    else `capacity -= capacity_diff`
+//
+// On a high level, `capacity` trends towards `baseline` by `capacity_diff` but does not “cross” it.
+func (k Keeper) UpdateCapacityForLimitParams(
+	ctx sdk.Context,
+	limitParams types.LimitParams,
+) {
+	supply := k.bankKeeper.GetSupply(ctx, limitParams.Denom)
+
+	capacityList := k.GetDenomCapacity(ctx, limitParams.Denom).CapacityList
+	if len(capacityList) != len(limitParams.Limiters) {
+		// This violates an invariant. Since this is in the `EndBlocker`, we log an error instead of panicking.
+		k.Logger(ctx).Error(
+			fmt.Sprintf(
+				"denom (%s) capacity list length (%v) != limiters length (%v); skipping capacity update",
+				limitParams.Denom,
+				len(capacityList),
+				len(limitParams.Limiters),
+			),
+		)
+		return
+	}
+
+	// Convert timestamps to milliseconds for algebraic operations.
+	blockTimeMilli := ctx.BlockTime().UnixMilli()
+	prevBlockInfo := k.blockTimeKeeper.GetPreviousBlockInfo(ctx)
+	prevBlockTimeMilli := prevBlockInfo.Timestamp.UnixMilli()
+	timeSinceLastBlockMilli := new(big.Int).Sub(
+		big.NewInt(blockTimeMilli),
+		big.NewInt(prevBlockTimeMilli),
+	)
```

Comment:
nit: this could be left as `int64` instead of `Big.Int`. Later can use `new(big.Rat).SetInt64(...)`